Lib/importlib/resources.py

```diff
@@ -268,38 +237,85 @@ def contents(package: Package) -> Iterator[str]:
             not package.__spec__.has_location):
         return []
     package_directory = Path(package.__spec__.origin).parent
-    try:
-        yield from os.listdir(str(package_directory))
-    except (NotADirectoryError, FileNotFoundError):
-        # The package is probably in a zip file.
-        archive_path = getattr(package.__spec__.loader, 'archive', None)
-        if archive_path is None:
-            raise
-        relpath = package_directory.relative_to(archive_path)
-        with ZipFile(archive_path) as zf:
-            toc = zf.namelist()
-        subdirs_seen = set()                        # type: Set
-        for filename in toc:
-            path = Path(filename)
-            # Strip off any path component parts that are in common with the
-            # package directory, relative to the zip archive's file system
-            # path.  This gives us all the parts that live under the named
-            # package inside the zip file.  If the length of these subparts is
-            # exactly 1, then it is situated inside the package.  The resulting
-            # length will be 0 if it's above the package, and it will be
-            # greater than 1 if it lives in a subdirectory of the package
-            # directory.
-            #
-            # However, since directories themselves don't appear in the zip
-            # archive as a separate entry, we need to return the first path
-            # component for any case that has > 1 subparts -- but only once!
-            if path.parts[:len(relpath.parts)] != relpath.parts:
+    yield from os.listdir(str(package_directory))
+
+
+# Private implementation of ResourceReader and get_resource_reader() for
+# zipimport.  Don't use these directly!  We're implementing these in Python
+# because 1) it's easier, 2) zipimport will likely get rewritten in Python
+# itself at some point, so doing this all in C would just be a waste of
+# effort.
+
+class _ZipImportResourceReader(resources_abc.ResourceReader):
+    """Private class used to support ZipImport.get_resource_reader().
+
+    This class is allowed to reference all the innards and private parts of
+    the zipimporter.
+    """
+
+    def __init__(self, zipimporter, fullname):
+        self.zipimporter = zipimporter
+        self.fullname = fullname
+
+    def open_resource(self, resource):
+        path = f'{self.fullname}/{resource}'
```

Comment:
Does zip importer require the forward slash? Or does it expect `os.sep`? (You're using pathlib below on what looks like it's probably the same style of path, but without digging further I'm not 100% sure.)

Reply:
I'm not 100% sure either, but AFAICT, the paths inside the zip all have [forward slashes](https://web.archive.org/web/20070928174718/http://www.pkware.com/documents/casestudies/APPNOTE.TXT).

I suppose that could pose a problem with using pathlib on Windows, but given Appveyor passing, I think we can trust that it isn't a problem in practice, right?

Reply:
There is `pathlib.PurePath.as_posix()` to guarantee the forward slashes.

Reply:
Shouldn't `os.fspath()` be called on `resource`? And is there any worry of people accidentally passing in a subdirectory since there is no argument check?

Reply:
I don't *think* we need to do those things since people will be accessing these through the higher level API (I.e. `importlib.resources`).  I think the validation done at that level means we can trust what gets passed to us.  I also don't think we need to force forward slashes in the argument passed to `self.zipimporter.get_data()` since we're using forward slash in the f-string and zips only support forward slashes in their internal file names.